go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -271,29 +274,68 @@ func (c *Controller) syncServices(key string) error {
 
 	vips := collectServiceVIPs(service)
 	if svcNeedsIdling(service.Annotations) && !util.HasValidEndpoint(service, endpointSlices) {
+		// addServiceToIdlingBalancer adds the vips to service tracker
 		err = c.addServiceToIdlingBalancer(vips, service)
 		if err != nil {
 			c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToAddToIdlingBalancer",
 				"Error trying to add to Idling LoadBalancer for Service %s/%s: %v", name, namespace, err)
 			return err
 		}
-		vipsTracked = vipsTracked.Delete(vips.List()...)
-		err = deleteVIPsFromNonIdlingOVNBalancers(vips, name, namespace)
+		// create a temporary map to convert sets.String to map for passing to functions
+		tmpVIPs := make(map[string]string)
+		for v := range vips {
+			delete(vipsTracked, v)
+			tmpVIPs[v] = ""
+		}
+		needsCleanUp := false
+		for vipProtocol := range vips {
+			vip, _ := splitVirtualIPKey(vipProtocol)
+			if c.serviceTracker.getLoadBalancer(name, namespace, vip) != loadbalancer.IdlingLoadBalancer {
+				needsCleanUp = true
+				break
+			}
+		}
+		if needsCleanUp {
+			err = deleteVIPsFromNonIdlingOVNBalancers(tmpVIPs, name, namespace)
+			if err != nil {
+				c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToDeleteOVNLoadBalancer",
+					"Error trying to delete the OVN LoadBalancer while setting up Idling for Service %s/%s: %v", name, namespace, err)
+				return err
+			}
+		}
+		// at this point we have processed all vips we've found in the service
+		// so the remaining ones that we had in the vipsTracked variable should be deleted
+		// We remove them from OVN and from the tracker
+		err = deleteVIPsFromAllOVNBalancers(vipsTracked, name, namespace)
 		if err != nil {
 			c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToDeleteOVNLoadBalancer",
-				"Error trying to delete the OVN LoadBalancer while setting up Idling for Service %s/%s: %v", name, namespace, err)
+				"Error trying to delete the OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
 			return err
 		}
 		// Delete the Service VIP from the Service Tracker
 		c.serviceTracker.deleteServiceVIPs(name, namespace, vipsTracked)
 		return nil
 	}
 
-	err = deleteVIPsFromIdlingBalancer(vips, name, namespace)
-	if err != nil {
-		c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToDeleteOVNLoadBalancer",
-			"Error trying to delete the idling OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
-		return err
+	needsCleanUp := false
+	tmpVIPs := make(map[string]string)
+	for vipProtocol := range vips {
+		vip, _ := splitVirtualIPKey(vipProtocol)
+		foundLb := c.serviceTracker.getLoadBalancer(name, namespace, vip)
+		// if vip was on an idling load balancer and we get here, we know we need to clean up
+		// if the value is empty (unknown what load balancer the vip may have existed on) also need to clean up
+		if len(foundLb) == 0 || foundLb == loadbalancer.IdlingLoadBalancer {
+			needsCleanUp = true
+		}
+		tmpVIPs[vipProtocol] = ""
```

Comment:
Same as above, can you put this under the condition? This will make it possible to get rid of `needsCleanUp` (and will save some unnecessary calls to deleteVIPsFromIdlingBalancer

Reply:
this is what 12 am code looks like :)